// sweet2/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.helpers = exports.createTemplateRenderer = exports.createMarkdownRenderer = exports.start = exports.paginate = exports.groupDocuments = exports.orderDocuments = exports.filterDocuments = exports.savePages = exports.generatePages = exports.loadSourceFiles = exports.loadConfig = undefined;

var _core = require('./core');

Object.defineProperty(exports, 'loadConfig', {
	enumerable: true,
	get: function get() {
		return _core.loadConfig;
	}
});
Object.defineProperty(exports, 'loadSourceFiles', {
	enumerable: true,
	get: function get() {
		return _core.loadSourceFiles;
	}
});
Object.defineProperty(exports, 'generatePages', {
	enumerable: true,
	get: function get() {
		return _core.generatePages;
	}
});
Object.defineProperty(exports, 'savePages', {
	enumerable: true,
	get: function get() {
		return _core.savePages;
	}
});
Object.defineProperty(exports, 'filterDocuments', {
	enumerable: true,
	get: function get() {
		return _core.filterDocuments;
	}
});
Object.defineProperty(exports, 'orderDocuments', {
	enumerable: true,
	get: function get() {
		return _core.orderDocuments;
	}
});
Object.defineProperty(exports, 'groupDocuments', {
	enumerable: true,
	get: function get() {
		return _core.groupDocuments;
	}
});
Object.defineProperty(exports, 'paginate', {
	enumerable: true,
	get: function get() {
		return _core.paginate;
	}
});

var _util = require('./util');

Object.defineProperty(exports, 'start', {
	enumerable: true,
	get: function get() {
		return _util.start;
	}
});

var _markdown = require('./renderers/markdown');

Object.defineProperty(exports, 'createMarkdownRenderer', {
	enumerable: true,
	get: function get() {
		return _markdown.default;
	}
});

var _template = require('./renderers/template');

Object.defineProperty(exports, 'createTemplateRenderer', {
	enumerable: true,
	get: function get() {
		return _template.default;
	}
});

var _helpers = require('../lib/helpers');

var helpers = _interopRequireWildcard(_helpers);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

exports.helpers = helpers;

// Default helpers
